package-best-practices.md

```diff
@@ -0,0 +1,46 @@
+
+There are hundreds of packages on the dbt Package Hub already, and over time we have identified a handful of common mistakes that new package authors make as they take a self-contained package and prepare it for publication. We have listed these below to help you and your users get the best experience.
+
+Although the dbt Labs team completes a cursory review of new packages before adding them to the Package Hub, ultimately you are responsible for your package's performance.
+
+To avoid ambiguity, the bullets below are pretty formal, but we are a friendly bunch! If you need help preparing a package for deployment to the dbt Package Hub, reach out in [#package-ecosystem](https://getdbt.slack.com/archives/CU4MRJ7QB/) in the [dbt Community Slack](https://getdbt.com/community).
+
+---
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).
+
+### First run experience
+- Packages SHOULD contain a README file which explains how to get started with the package and customise its behaviour.
+- The documentation SHOULD indicate which data warehouses are expected to work with this package.
+
+### Customisability 
+- Packages MUST NOT hard-code table references, and MUST use `ref` or `source` instead.
+- Packages MAY provide a mechanism (such as variables) to enable/disable certain subsets of functionality.
+- Packages for data transformation:
+    - SHOULD provide a mechanism (such as variables) to customise the location of source tables.
+    - SHOULD NOT assume database/schema names in sources. 
+    - MAY assume table names, particularly if the package was built to support tables created by a known tool.
+
+### Dependencies
+#### Dependencies on dbt Core
+- Packages requiring a minimum version of `dbt-core` MUST declare it in the `require-dbt-version` property of `dbt_project.yml`.
+- Packages requiring a minimum version of `dbt-core` SHOULD allow all subsequent minor and patch releases of that major version. 
+    - For example, a package which depends on functionality added in dbt Core 1.2 SHOULD set its `require-dbt-version` property to `[">=1.2.0", "<2.0.0"]`.
+#### Dependencies on other packages defined in `packages.yml`:
+- Packages SHOULD import their dependencies from the dbt Package Hub when available, as opposed to a `git` installation.
+- Packages SHOULD NOT pin to a patch version of their imported package unless they are aware of an incompatibility.
+- Packages importing another package which has reached major version one or later SHOULD allow all subsequent minor and patch releases of that major version. 
+- Packages importing another package which is major version zero MAY pin to the current minor version only.
```

Comment:
I think the last 3 bullet points can all be replaced with simple guidance:

```suggestion
- The widest version range possible SHOULD be specified to give the dependency resolver the appropriate latitude.
```

My guess is that after lots of discussion exploring many scenarios, ☝️ will still hold as an essential summary, a [cataclysm sentence](https://radiolab.org/episodes/cataclysm-sentence) that "contains the most information in the fewest words".

The three bullet points make assumptions about the versioning scheme being used by the imported package which _might_ be strict [SemVer 2.0](https://semver.org/spec/v2.0.0.html), but just as likely might be a variant or something else entirely.

Reply:
>The three bullet points make assumptions about the versioning scheme being used by the imported package which might be strict [SemVer 2.0](https://semver.org/spec/v2.0.0.html), but just as likely might be a variant or something else entirely.

Could we change that `SHOULD` from Releases and Updates to a `MUST`? 

Those bullets are written in the same spirit as when you go to a fancy hotel and there's a _No Elephants Allowed In the Ballroom_ sign - you know they didn't put it up until they had to. 

The first one (no patches) is the most common and catastrophic failure state I see; I'm not wed to the others but they seemed like the obvious next question: "If I'm not allowed to pin to a patch, what _do_ you recommend?"

What do you think of something like this?
> Packages SHOULD specify the widest possible range of supported versions, to minimise issues in dependency resolution. In particular, packages SHOULD NOT pin to a patch version of their imported package unless they are aware of an incompatibility.

Reply:
> Packages SHOULD specify the widest possible range of supported versions, to minimise issues in dependency resolution. In particular, packages SHOULD NOT pin to a patch version of their imported package unless they are aware of an incompatibility.

🎯 looks perfect

> Could we change that `SHOULD` from Releases and Updates to a `MUST`?

I'm _guessing_ that we can't say "Packages MUST follow the guidance of Semantic Versioning 2.0.0" because we don't have any auditing mechanism -- there is no easy way to validate to what degree a package is following it or not. We _can_ validate the version string with a regex, but we _can't_ reasonably determine if they are following the SemVer guidance or not. They are on the honor system for this one.

Reply:
Fair!